strawberry/schema/execute.py

```diff
@@ -256,3 +259,135 @@ def execute_sync(
+    The graphql-core subscribe function returns either an ExecutionResult or an
+    AsyncGenerator[ExecutionResult, None].  The former is returned in case of an error
+    during parsing or validation.
+    Because we need to maintain execution context, we cannot return an
+    async generator, we must _be_ an async generator.  So we yield a
+    (bool, ExecutionResult) tuple, where the bool indicates whether the result is an
+    potentially multiple execution result or a single result.
+    A False value indicates an single result, most likely an intial
+    failure (and no more values will be yielded) whereas a True value indicates a
+    successful subscription, and more values may be yielded.
+    """
+
+    extensions_runner = SchemaExtensionsRunner(
+        execution_context=execution_context,
+        extensions=list(extensions),
+    )
+
+    # unlike execute(), the entire operation, including the results hooks,
+    # is run within the operation() hook.
+    async with extensions_runner.operation():
+        # Note: In graphql-core the schema would be validated here but in
+        # Strawberry we are validating it at initialisation time instead
+        assert execution_context.query is not None
+
+        async with extensions_runner.parsing():
+            try:
+                if not execution_context.graphql_document:
+                    execution_context.graphql_document = parse_document(
+                        execution_context.query, **execution_context.parse_options
+                    )
+
+            except GraphQLError as error:
+                execution_context.errors = [error]
+                process_errors([error], execution_context)
+                yield False, ExecutionResult(
+                    data=None,
+                    errors=[error],
+                    extensions=await extensions_runner.get_extensions_results(),
+                )
+                # the generator is usually closed here, so the following is not
+                # reached
+                return  # pragma: no cover
+
+            except Exception as error:  # pragma: no cover
+                error = GraphQLError(str(error), original_error=error)
+                execution_context.errors = [error]
+                process_errors([error], execution_context)
+                yield False, ExecutionResult(
+                    data=None,
+                    errors=[error],
+                    extensions=await extensions_runner.get_extensions_results(),
+                )
+                return  # pragma: no cover
+
+        async with extensions_runner.validation():
+            _run_validation(execution_context)
+            if execution_context.errors:
+                process_errors(execution_context.errors, execution_context)
+                yield False, ExecutionResult(data=None, errors=execution_context.errors)
+                return  # pragma: no cover
+
+        async def process_result(result: GraphQLExecutionResult):
```

Comment:
Why is that defined here, on every execution?

Reply:
Its a closure.  Avoids writing a compicated external function and passing arguments, as well as scoping functionality.
Closures aren't actually _defined_ when run, merely instantiated with bindings.  I guess it is down to style, I very much favor closures for their conciseness and locality.  Happy to change it if you like.

Reply:
I think it's a weak point for performance, I would be happy if we would merge #2810 here or vice-versa. There I added a base class for async execution and it is done for subscription and async execution via   https://github.com/nrbnlulu/strawberry/blob/support_extensions_on_subscriptions/strawberry/schema/execute.py#LL225C15-L225C39

Reply:
I wasn't aware of #2810, I'll look at that.

I just want to add, as a general comment, that I think you will find that performance is not affected.
First of all, instantiate a closure a drop in the ocean of all the stuff that goes on to actually start executing the subscription.
Secondly, all of the execution time of a subscription actually happens parsing and processing the multiple results.  Setting up of an subscription is not performance critical.

Finally, you suggested using `contextlib.suppress`, and that involves not only instantiating a class on every query, but also invoking two methods on it.  There is quite a bit of machinery involved in a a `contextlib.contextmanager`.

I understand when people are concerned about performance, but one always needs to look at that in context.

Reply:
> I understand when people are concerned about performance, but one always needs to look at that in context.

Thanks. ig it is just something that bugs me... unless it is in a decorator I won't do it...
Sorry if i'm being too strict.

Reply:
Sure, its moved out of the function in the latest commit.